File: Trees/bst.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import random
import sys

# shamelessly stolen from SO :-P
class TreeNode:

    def __init__(self, value):
        self.left = None
        self.right = None
        self.data = value

class Tree:

    def __init__(self):
        self.root = None

    def addNode(self, node, value):
        if node is None:
            self.root = TreeNode(value)
        else:
            if (value < node.data):
                if node.left is None:
                    node.left = TreeNode(value)
                else:
                    self.addNode(node.left, value)
            else:
                if node.right is None:
                    node.right = TreeNode(value)
                else:
                    self.addNode(node.right, value)

    def printInOrder(self, node):
        if node is not None:
            self.printInOrder(node.left)
            print(node.data)
            self.printInOrder(node.right)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("nodes", type=int,
                    help="how many nodes to generate", default=100)
    args = parser.parse_args()
    bst = Tree()
    for i in range(0, args.nodes):
        bst.addNode(bst.root, random.randint(0,2048))
    bst.printInOrder(bst.root)

if __name__ == '__main__':
    sys.exit(main())

File: golang/Trees/bst.go
package main

import "fmt"
import "math/rand"

type Node struct {
	key   int
	left  *Node
	right *Node
}

func (root *Node) insert(newNode *Node) {
	if root.key == newNode.key {
		return
	}
	if root.key < newNode.key {
		if root.right == nil {
			root.right = newNode
		} else {
			root.right.insert(newNode)
		}
	} else {
		if root.left == nil {
			root.left = newNode
		} else {
			root.left.insert(newNode)
		}
	}
}

func (root *Node) inOrder(fn func(n *Node)) {

	if root.left != nil {
		root.left.inOrder(fn)
	}

	fn(root)

	if root.right != nil {
		root.right.inOrder(fn)
	}

}

type Bst struct {
	root   *Node
	length int
}

func (bst *Bst) insert(value int) {
	node := &Node{key: value}
	bst.length++

	if bst.root == nil {
		bst.root = node
	} else {
		bst.root.insert(node)
	}
}

func (bst *Bst) inorderPrint() {
	bst.root.inOrder(func(node *Node) {
		fmt.Println(node.key)
	})
}
func main() {
	tree := new(Bst)
	for i := 0; i < 0x20; i++ {
		tree.insert(rand.Int())
	}
	tree.inorderPrint()
}
